### ISSUE TITLE
[REST_CherryPy] Is it possible for session cookies to specify their domain?

### ISSUE THREAD

**Issue by NickDubelman** [saltstack/salt#38141]
### Description of Issue/Question
I'm considering the case where salt-api is running on a different server than a deployed app that makes use of the API (likely a common use case, I imagine?) 

Currently, since the token does not specify a domain, the deployed app can not consume the cookie contained in the Set-Cookie header of the response. 

Is it feasible for the API to use the hostname of the machine it's running on and include this as the domain when creating/returning the cookie? 

As I understand it, this would allow an app hosted at foo.bar.com to query the API hosted at baz.bar.com.

Thanks much

**Comment by whiteinge** [comment]
What kind of client are you using? If this is a browser client rest_cherrypy has support for both simple CORS and the full preflight request CORS. The `Access-Control-Allow-Credentials` header is set so cookies from the remote domain should indeed be automatically sent and received. (Same-origin will still restrict read-access from JavaScript but you shouldn't need to read them directly since the browser will automatically send and receive them.)

All that said, I honestly don't know how well vetted that CORS support is so if you run into trouble please include your steps and what trouble you're seeing.

**Comment by NickDubelman** [comment]
I've noticed that when I omit credentials from a request and instead try to pass an X-Auth-Token, since this is a custom header, the client will first issue an OPTIONS request to perform the preflight request. This OPTIONS request always returns 401 Unauthorized for me. 

I also believe there is a security vulnerability-- I believe rest_cherrypy is vulnerable to CSRF attacks. As I understand it, a CSRF attack exploits the fact that browsers save cookies for the domain in which a given request occurred.  Meaning that if I am already authenticated with the server running my salt-api, someone could send me a link that makes a malicious request to this API, and since I am already authenticated, the request will succeed. I have no experience with combatting such vulnerabilities, but I am fairly positive that this is indeed a vulnerability.

**Comment by whiteinge** [comment]
I'll take a look at that OPTIONS request. It sounds like auth may need to be disabled there.

rest_cherrypy provides two methods to track authentication, one by cookies and one by the custom header. If you build a browser-based app that relies on cookies for auth then it will be vulnerable to CSRF but if you use JavaScript to perform the login by extracting and re-sending the custom header you will be safe (it becomes a de-facto CSRF token in that workflow).

Thanks for mentioning this! We haven't shipped a default web app with rest_cherrypy before so the docs don't really spell that out or make a recommendation either way. Plus the default configuration makes it kind of easy to shoot yourself in the foot. We are shipping a default web app as a POC in the next Salt release so we should do better here. I'll get xsrf support added. I think it would also be good to be able to turn off cookie as a config flag.

**Comment by whiteinge** [comment]
@NickDubelman sorry for the delay. Are you in a position to test #38224 to see if that does the trick for CORS?

**Comment by NickDubelman** [comment]
Yeah definitely, sorry for the delay. Do you mind holding my hand for that process-- should I just replace rest_cherrypy/app.py in the src and delete the .pyc?

**Comment by whiteinge** [comment]
Yeah, that should do the trick. :-)

**Comment by NickDubelman** [comment]
I'm still getting 401 Unauthorized on the Preflight OPTIONS request. 

This is the (successful) login request.
```
async login(username, password){
  let request = {
    method: 'POST',
    headers: {
      'Accept': 'application/json',
    },
    body: JSON.stringify({
      username: username,
      password: password,
      eauth: 'pam',
    })
  }
  try{
    let response = await fetch('someurl:8000/login', request)
    console.log(response.status)
    if (response.status < 200 || response.status >= 300) {
      throw({message: "HTTP Error"})
    }
    else{
      let json = await response.json()
      let ret = json.return[0]
      let token = ret.token
      let {expire} = ret
      localStorage.setItem("token", JSON.stringify({token, expire}))
      console.log(JSON.parse(localStorage.token))
      this.setState({loggedIn: true}, this.getMinions)
    }
  } catch(err){
    alert("Invalid Credentials")
    console.log(err)
  }
}
```

And this is the unsuccessful minions request. 
```
async getMinions(){
  let request = {
    headers: {
      'Accept': 'application/json',
      'X-Auth-Token': JSON.parse(localStorage.token).token
    }
  }
  try{
    let response = await fetch('someurl:8000/minions', request)
    let json = await response.json()
    let minions = objToArray(json.return[0])
    this.setState({minions})
  } catch(err){
    console.log(err)
  }
}
```

This is the unsuccessful request: 
<img width="758" alt="screen shot 2016-12-13 at 4 19 19 pm" src="https://cloud.githubusercontent.com/assets/16615307/21165025/8b95fa2a-c151-11e6-9288-be6174443ba9.png">

While it isn't related to the aforementioned issue of receiving the 401 Unauthorized on the Preflight request, notice that the request is still returning (and creating) a session cookie, _even though I did not specify to include credentials on the request_. This means that your suggested strategy of "opting out" of cookies will not work _because even in the case where you opt out the client will still create the cookie and therefore CSRF vulnerability can be exploited. I liked your idea of using a config flag to opt out of cookies entirely.

I've been working on a web app to use as a playground for interacting with Salt, it will take me a while to remove some the company specific part of my code but I can share that as a repo if youd like. It's a react app so it will require npm for you to run, but it would give you an instant javascript playground to test salt-api in a browser context. However, you mentioned that you guys are getting close to shipping a proof of concept web app, which could fulfill the same niche.

**Comment by whiteinge** [comment]
Hm. That change should've allowed unauthed OPTIONS requests for all
endpoints. I must have missed something. Sorry for the churn and
thanks for testing it for me! It's been a hectic week so far but I can
make time tomorrow for a closer look.

**Comment by whiteinge** [comment]
I got a little buried in the run-up to the holidays. Still digging out
but this is still on my radar.

**Comment by Ch3LL** [comment]
@whiteinge any update on this we would like to get it into nitrogen if possible

**Comment by whiteinge** [comment]
@NickDubelman if this is still relevant to you, the above PR fixes the OPTIONS request (verified) and adds a setting to disable session endpoints altogether. Thanks for all your help with this one!

**Comment by cachedout** [comment]
Closed via #41352
